### PR TITLE
[4.0] Make categories use an interface

### PR DESCRIPTION
**Pull request by wilsonge**
- No hardcoding namespaces in the provider - grab them from the db so there's a single point of truth
- Use a feature interface for categories with an exception interface

### CODE REVIEW

**Comment by ghost** [comment]
Changed Title to be clear its about 4.0.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20170">issues.joomla.org/tracker/joomla-cms/20170</a>.</sub>

File: administrator/components/com_content/services/provider.php [hunk 1]
```diff
@@ -56,13 +65,16 @@ public function register(Container $container)
 		$container->set(Categories::class, ['' => new Category]);
 		$container->set(AssociationExtensionInterface::class, new AssociationsHelper);
 
-		$factory = new MVCFactoryFactory('\\Joomla\\Component\\Content');
+		$factory = new MVCFactoryFactory(ComponentHelper::getComponent($this->extension_name)->namespace);
```
Comment (laoneo) [inline administrator/components/com_content/services/provider.php:68]: A service provider should have as less external dependencies which can't be leaded trough the container as possible, especially not static ones. This line makes the service provider dependent on an external environment, in this case it even needs a fully installed CMS.
The service provider is the point of truth and it is absolutely ok to hardcode here the root namespace. Ideally you can even make classes in your own component which are the factories and then you are not dependent on any namespace. This here are mainly helper classes to fulfill a default case.

**Comment by laoneo** [comment]
While rewriting #19843 (which is a similar pr) I'v discovered one issue you were raising in the last meeting as well. It looks awkward when an extension is implementing `AssociationAwareInterface` and at the same time a `AssociationsNotImplementedException` exception can be thrown. I know code wise it is correct, but from a logical point of view it is questionable if this is really the way to go.

That's why I made #20127 which showcases a generic service locator. The big drawback on that approach is that we do not have typed return values anymore.

File: administrator/components/com_content/services/provider.php (hunk 1 above)
Comment (wilsonge) [reply administrator/components/com_content/services/provider.php:68]: Our container here isn't really a container - it's a service locator and yes an extension for Joomla will require a fully installed CMS - I don't see this as a problem. We should have the namespace in a single place - and currently that place is in the extensions XML file.

**Comment by wilsonge** [comment]
> It looks awkward when an extension is implementing AssociationAwareInterface and at the same time a AssociationsNotImplementedException

I agree - it's just the price of supporting the legacy implementation. I discussed this at length with Nic in Holland and the eventual outcome was it was less painful to have the exception than to loose the typehint (as the exception can easily be removed in j5 or whatever - but modifying an interface is a hard b/c break)

File: administrator/components/com_content/services/provider.php (hunk 1 above)
Comment (laoneo) [reply administrator/components/com_content/services/provider.php:68]: This setting is not used anywhere and not needed with the service provider approach. It should be removed. It adds some not needed complexity and dependency.

**Comment by laoneo** [comment]
Don't think so it is a BC break problem. It reveals the problems when one extension is supporting associations and another not and both are using the default Component service provider.

**Comment by laoneo** [comment]
Made an alternative in #20176 which introduces a component class which defines the services it provides.

**Comment by wilsonge** [comment]
Closing as I've merged https://github.com/joomla/joomla-cms/pull/20217

**Pull request closed without merging**